pkg/apis/core/validation/validation.go

```diff
@@ -5405,3 +5406,78 @@ func ValidateProcMountType(fldPath *field.Path, procMountType core.ProcMountType
 		return field.NotSupported(fldPath, procMountType, []string{string(core.DefaultProcMount), string(core.UnmaskedProcMount)})
 	}
 }
+
+var (
+	supportedScheduleActions = sets.NewString(string(core.DoNotSchedule), string(core.ScheduleAnyway))
+)
+
+type spreadConstraintPair struct {
+	topologyKey    string
+	scheduleAction core.UnsatisfiableConstraintAction
+}
+
+// validateTopologySpreadConstraints validates given TopologySpreadConstraints.
+func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstraint, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	var existingConstraintPairs []spreadConstraintPair
+	for i, constraint := range constraints {
+		subFldPath := fldPath.Index(i)
+		if err := ValidateMaxSkew(subFldPath.Child("maxSkew"), constraint.MaxSkew); err != nil {
+			allErrs = append(allErrs, err)
+		}
+		if err := ValidateTopologyKey(subFldPath.Child("topologyKey"), constraint.TopologyKey); err != nil {
+			allErrs = append(allErrs, err)
+		}
+		if err := ValidateScheduleAction(subFldPath.Child("whenUnsatisfiable"), constraint.WhenUnsatisfiable); err != nil {
+			allErrs = append(allErrs, err)
+		}
+		// tuple {topologyKey, whenUnsatisfiable} denotes one kind of spread constraint
+		pair := spreadConstraintPair{
+			topologyKey:    constraint.TopologyKey,
+			scheduleAction: constraint.WhenUnsatisfiable,
+		}
+		if err := ValidateSpreadConstraintPair(subFldPath.Child("{topologyKey, whenUnsatisfiable}"), pair, existingConstraintPairs); err != nil {
+			allErrs = append(allErrs, err)
+		} else {
+			existingConstraintPairs = append(existingConstraintPairs, pair)
+		}
+	}
+
+	return allErrs
+}
+
+// ValidateMaxSkew tests that the argument is a valid MaxSkew.
+func ValidateMaxSkew(fldPath *field.Path, maxSkew int32) *field.Error {
+	if maxSkew <= 0 {
+		return field.Invalid(fldPath, maxSkew, isNotPositiveErrorMsg)
+	}
+	return nil
+}
+
+// ValidateTopologyKey tests that the argument is a valid TopologyKey.
+func ValidateTopologyKey(fldPath *field.Path, topologyKey string) *field.Error {
+	if len(topologyKey) == 0 {
+		return field.Required(fldPath, "can not be empty")
+	}
+	return nil
+}
+
+// ValidateScheduleAction tests that the argument is a valid UnsatisfiableConstraintAction.
+func ValidateScheduleAction(fldPath *field.Path, action core.UnsatisfiableConstraintAction) *field.Error {
+	if !supportedScheduleActions.Has(string(action)) {
+		return field.NotSupported(fldPath, action, supportedScheduleActions.List())
+	}
+	return nil
+}
+
+// ValidateSpreadConstraintPair tests that if `pair` exists in `existingConstraintPairs`.
+func ValidateSpreadConstraintPair(fldPath *field.Path, pair spreadConstraintPair, existingConstraintPairs []spreadConstraintPair) *field.Error {
+	for _, existingPair := range existingConstraintPairs {
+		if pair.topologyKey == existingPair.topologyKey &&
+			pair.scheduleAction == existingPair.scheduleAction {
```

Comment:
i am not following , it seems you are suggesting we can only have a unique pair so topoKey1,scheduleAction1 and topoKey1,scheduleAction2 are valid pairs ? What would that mean in implementation ? I think if we have topology key repeated that is an error , and  we dont need to check for schedule action @Huang-Wei

Reply:
Two constraints with same {topologyKey, scheduleAction} should be mergeable:

- greater maxSkew should be respected
- labelSelectors are also mergeable

So for the sake of calculation efficiency and conciseness of API, I do believe we should only allow different constraints with different {topologyKey, scheduleAction} tuples.

Reply:
IMO, it makes it simpler to reason about if we don't allow duplicate topologyKey. We should explicitly flag this as an error rather than silently behind the scenes merge the constraints and leave the user with a behavior he/she didnt expect. If the user wanted something that needs to be merged, he can specify a merged constraint. I see this as more of better usability

Reply:
The merge I mean is patch merging in API side, not that 2 constraints with same topokey+scheduleAction in an API request will be merged - which will fail.

Reply:
Duplicate key makes lots of sense if you prefer skews below X, and can't tolerate skews above Y, X < Y.

Reply:
interesting Idea, thanks for bringing that up @lavalamp . Its too early to say we want this, but lets say we want to implement this, then we have two options:-
```
**option1:** 
{ 
   topologyKey: topo1
   maxSkew: 3
} 
{ 
   topologyKey: topo1
   maxSkew: 5
} 


**option2:** 
{ 
   topologyKey: topo1
   skewRange: {
    min: 3
    max: 5
  }  
```

I think option 2 is much more clear and user friendly. WDYT @lavalamp @bsalamat @Huang-Wei ?


@Huang-Wei the UT you have says if you find duplicate tuple(topoKey and scheduleAction), you will fail. That means your validation will pass for tuples with same topoKey and different scheduleActions. 

I am suggesting if we decide to not implement the idea by @lavalamp , then we should not allow tuples with same topoKey and different scheduleActions. It will be difficult to document what that behavior means . We should simplify this and say two constraints with same topoKey dont make any sense and we should error out .

Reply:
@krmayankk I'm not sure if we're talking about the same thing. Both @lavalamp and I are just talking about the following combination can be possible (where X<Y):

1. topologyKey=zone, scheduleAction=ScheduleAnyway, maxSkew=X - it's a **soft** requirement, i.e. it's a good-to-have
1. topologyKey=zone, scheduleAction=DoNotSchedule, maxSkew=Y - it's a **hard** requirement, i.e. it's a must-have

In above case, the 2nd constraint MUST be obeyed, and the 1st constraint is just _preferred_.

---

The `skewRange` thing you raised is confusing b/c when it comes to term "range", it reminds me of the formula "min < actual skew < max", which is absolutely not the same thing "even pods spread" is talking about.

Reply:
I see your point. Is your subsequent PR's implementing this logic already ?

The range was just an example,We could call it something else, but the point is , the user should not have to repeat the topologySpreadConstraint, if they just need to specify a soft and a hard requirement with different skew.  So it could look like this:
```
   topologyKey=zone
   softSkew: X
   hardSkew: Y
```
One of softSkew or hardSkew must be specified. I think this is easier and more user friendly representation then 
```
{
    topologyKey=zone
    scheduleAction=ScheduleAnyway
    maxSkew=Y
},
{
    topologyKey=zone
    scheduleAction= DoNotSchedule
    maxSkew=Y
}
```
Its an alpha so we can get feedback from the community, but I have a hard time thinking why someone would choose the second option over the first one .

Reply:
Please leave this the way it is. Putting the action in fields (whether min/max or soft/hard) like this makes it much harder to add additional constraint types in the future.

Reply:
@krmayankk I see your point. But I agree more with @lavalamp - to make the options standalone (instead of mixing them in the beginning) make it clear and extensible.